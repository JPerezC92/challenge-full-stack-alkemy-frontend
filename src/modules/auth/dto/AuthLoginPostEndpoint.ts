import {
  JsendError,
  JsendFail,
  JsendSuccess,
} from "src/modules/shared/service/JsendResponse";
import { UserEndpoint } from "./UserEndpoint";

export type AuthLoginPostEndpoint =
  | JsendSuccess<{
      accessToken: string;
      refreshToken: string;
      user: UserEndpoint;
    }>
  | JsendError
  | JsendFail;
