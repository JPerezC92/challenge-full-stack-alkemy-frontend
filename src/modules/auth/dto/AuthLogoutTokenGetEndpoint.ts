import {
  JsendError,
  JsendFail,
  JsendSuccess,
} from "src/modules/shared/service/JsendResponse";

export type AuthLogoutTokenGetEndpoint = JsendSuccess | JsendError | JsendFail;
