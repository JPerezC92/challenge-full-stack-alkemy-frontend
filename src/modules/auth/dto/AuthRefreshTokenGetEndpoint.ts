import {
  JsendError,
  JsendFail,
  JsendSuccess,
} from "src/modules/shared/service/JsendResponse";
import { UserEndpoint } from "./UserEndpoint";

export type AuthRefreshTokenGetEndpoint =
  | JsendSuccess<{
      accessToken: string;
      refreshToken: string;
      user: UserEndpoint;
    }>
  | JsendError
  | JsendFail;
