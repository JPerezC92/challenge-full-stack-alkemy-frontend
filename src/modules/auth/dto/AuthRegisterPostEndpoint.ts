import {
  JsendError,
  JsendFail,
  JsendSuccess,
} from "src/modules/shared/service/JsendResponse";
import { UserEndpoint } from "./UserEndpoint";

export type AuthRegisterPostEndpoint =
  | JsendSuccess<{
      accessToken: string;
      refreshToken: string;
      user: UserEndpoint;
    }>
  | JsendError
  | JsendFail;
