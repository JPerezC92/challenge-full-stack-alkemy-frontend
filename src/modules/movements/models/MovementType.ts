export enum MovementType {
  INCOME = "income",
  EXPENSE = "expense",
}

export function isMovementType(
  movementType: any
): movementType is MovementType {
  return Object.values(MovementType).includes(movementType);
}
