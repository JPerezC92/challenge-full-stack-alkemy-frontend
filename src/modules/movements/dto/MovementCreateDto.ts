import { MovementType } from "../models/MovementType";

export interface MovementCreateDto {
  concept: string;
  amount: number;
  type: MovementType;
}
