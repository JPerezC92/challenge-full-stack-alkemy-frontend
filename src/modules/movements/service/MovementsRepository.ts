import { MovementEndpoint } from "src/modules/movements/dto/MovementEndpoint";
import { OrderType } from "src/modules/shared/models/OrderType";
import { MovementCreateDto } from "../dto/MovementCreateDto";

export interface MovementsRepository {
  query(query: {
    page?: number;
    limit?: number;
    order?: OrderType;
  }): Promise<MovementEndpoint[]>;
  create(movementCreateDto: MovementCreateDto): Promise<void>;
}
