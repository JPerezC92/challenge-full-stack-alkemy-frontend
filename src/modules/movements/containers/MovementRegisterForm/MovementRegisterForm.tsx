import React from "react";
import {
  MovementCreatedEventActionType,
  useMovementCreatedEventState,
} from "src/modules/movements/context/MovementCreatedEvent.context";
import { MovementCreate } from "src/modules/movements/dto/MovementCreateDto";
import { MovementType } from "src/modules/movements/models/MovementType";
import { MovementsRepository } from "src/modules/movements/service/MovementsRepository";
import { useCallableRequest } from "src/modules/shared/hooks/useCallableRequest";
import { useForm } from "src/modules/shared/hooks/useForm";
import { MyRepository } from "src/modules/shared/service/MyRepository";
import { capitalize } from "src/modules/shared/utils/capitalize";
import { getCurrentDateString } from "src/modules/shared/utils/getCurrentDateString";

type MovementRegisterFormProps = {
  movementsRepository: MyRepository<MovementsRepository>;
};

export const MovementRegisterForm: React.FC<MovementRegisterFormProps> = ({
  movementsRepository,
}) => {
  const { movementCreatedEventDispatch } = useMovementCreatedEventState();

  const createMovement = useCallableRequest(async ({ abortController }) => {
    const _movementsRepository = movementsRepository({ abortController });
    return async (movementCreateDto: MovementCreate) => {
      await _movementsRepository.create(movementCreateDto);

      movementCreatedEventDispatch({
        type: MovementCreatedEventActionType.MOVEMENT_CREATED,
        payload: movementCreateDto.type,
      });
    };
  }, []);

  const { formValues, names, ids, handleChange, handleSubmit } =
    useForm<MovementCreate>({
      initialValues: {
        concept: "",
        amount: "",
        type: "",
        date: getCurrentDateString(),
      },
      onSubmit: async (values, resetValues) => {
        await createMovement.execute({
          ...values,
          amount: Number(values.amount),
          type: values.type as MovementType,
        });
        resetValues();
      },
    });

  return (
    <>
      <form onSubmit={handleSubmit} role="form">
        <div>
          <label htmlFor={ids.concept}>{capitalize(names.concept)}: </label>
          <input
            id={ids.concept}
            name={names.concept}
            onChange={handleChange}
            type="text"
            value={formValues.concept}
            required
          />
        </div>

        <div>
          <label htmlFor={ids.amount}>{capitalize(names.amount)}</label>
          <input
            id={ids.amount}
            name={names.amount}
            onChange={handleChange}
            type="number"
            value={formValues.amount}
            required
          />
        </div>

        <div>
          <label htmlFor={ids.date}>{capitalize(names.date)}</label>
          <input
            id={ids.date}
            name={names.date}
            onChange={handleChange}
            type="date"
            value={formValues.date}
            required
          />
        </div>

        <div>
          <label htmlFor={ids.type}>{capitalize(names.type)}</label>
          <select
            id={ids.type}
            name={names.type}
            onChange={handleChange}
            value={formValues.type}
            required
          >
            <option key="" value="" disabled>
              Select a type
            </option>
            {Object.entries(MovementType).map(([key, value]) => (
              <option key={key} value={value}>
                {value}
              </option>
            ))}
          </select>
        </div>

        <button type="submit">Save</button>
      </form>
    </>
  );
};
