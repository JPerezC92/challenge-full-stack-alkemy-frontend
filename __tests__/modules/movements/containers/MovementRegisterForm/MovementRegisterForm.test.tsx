import { act, fireEvent, render } from "@testing-library/react";
import renderer from "react-test-renderer";
import { MovementRegisterForm } from "src/modules/movements/containers/MovementRegisterForm";
import { MovementCreateDto } from "src/modules/movements/dto/MovementCreateDto";
import { MovementType } from "src/modules/movements/models/MovementType";
import { MovementsRepository } from "src/modules/movements/service/MovementsRepository";
import * as useNodeMovementsRepository from "src/modules/movements/service/useNodeMovements.repository";

const movementsRepository: MovementsRepository = {
  create: jest.fn(),
  query: jest.fn(),
};

jest
  .spyOn(useNodeMovementsRepository, "useNodeMovementsRepository")
  .mockReturnValue(() => movementsRepository);

describe("MovementRegisterForm container", () => {
  test("should match the snapshot", async () => {
    const component = renderer
      .create(
        <MovementRegisterForm
          movementsRepository={useNodeMovementsRepository.useNodeMovementsRepository()}
        />
      )
      .toJSON();

    expect(component).toMatchSnapshot();
  });

  test("should call MovementsRepository.create with the correct parameters", async () => {
    const movementCreateDto: MovementCreateDto = {
      amount: 1000,
      concept: "concept",
      type: MovementType.INCOME,
    };

    const component = render(
      <MovementRegisterForm
        movementsRepository={useNodeMovementsRepository.useNodeMovementsRepository()}
      />
    );

    const form = component.getByRole("form");
    const conceptInput = component.getByLabelText("Concept");
    const amountInput = component.getByLabelText("Amount");
    const typeSelect = component.getByLabelText("Type");

    await act(async () => {
      fireEvent.change(conceptInput, {
        target: { value: movementCreateDto.concept },
      });
      fireEvent.change(amountInput, {
        target: { value: movementCreateDto.amount },
      });
      fireEvent.change(typeSelect, {
        target: { value: movementCreateDto.type },
      });

      fireEvent.submit(form);
    });

    expect(movementsRepository.create).toHaveBeenCalledTimes(1);
    expect(movementsRepository.create).toHaveBeenCalledWith(movementCreateDto);
  });
});
