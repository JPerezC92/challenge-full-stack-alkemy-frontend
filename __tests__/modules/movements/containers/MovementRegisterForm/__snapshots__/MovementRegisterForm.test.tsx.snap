// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovementRegisterForm container should match the snapshot 1`] = `
<form
  onSubmit={[Function]}
  role="form"
>
  <div>
    <label
      htmlFor=":r0:concept"
    >
      Concept
    </label>
    <input
      id=":r0:concept"
      name="concept"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
  </div>
  <div>
    <label
      htmlFor=":r0:amount"
    >
      Amount
    </label>
    <input
      id=":r0:amount"
      name="amount"
      onChange={[Function]}
      required={true}
      type="number"
      value=""
    />
  </div>
  <div>
    <label
      htmlFor=":r0:type"
    >
      Type
    </label>
    <select
      id=":r0:type"
      name="type"
      onChange={[Function]}
      required={true}
      value=""
    >
      <option
        disabled={true}
        value=""
      >
        Select a type
      </option>
      <option
        value="income"
      >
        income
      </option>
      <option
        value="expense"
      >
        expense
      </option>
    </select>
  </div>
  <button
    type="submit"
  >
    Save
  </button>
</form>
`;
